src/tools/lifter/bir_exp_liftingScript.sml

```diff
@@ -7,9 +7,305 @@ open bir_exp_memTheory bir_bool_expTheory
 open bir_nzcv_expTheory bir_interval_expTheory
 open bir_lifter_general_auxTheory
 open bir_extra_expsTheory
+open finite_mapTheory;
+open pred_setTheory;
 
 val _ = new_theory "bir_exp_lifting";
 
+(***********************************)
+(* memory map conversion functions *)
+(***********************************)
+
+
+val bir_mmap_w_w2n_def = Define `
+   (bir_mmap_w_w2n (mmap_w: 'a word |-> 'b word)): num |-> num =
+      FUN_FMAP (\x. w2n (FAPPLY mmap_w (n2w x)))
+               (IMAGE w2n (FDOM mmap_w))
+    `;
+
+val bir_mmap_n2w_def = Define `
+   (bir_mmap_n2w (mmap: num |-> num)) : 'a word |-> 'b word =
+      FUN_FMAP (\x. n2w (FAPPLY mmap (w2n x)))
+               (IMAGE n2w ((FDOM mmap) INTER {n | n < dimword (:'a)}))
+    `;
+
+val bir_load_mmap_w_def = Define `
+    bir_load_mmap_w (mmap_w: 'a word |-> 'b word) a =
+      n2w (bir_load_mmap (bir_mmap_w_w2n mmap_w) (w2n (a : 'a word))) : 'b word
+    `;
+
+val bir_load_mmap_w_alt_thm = store_thm("bir_load_mmap_w_alt_thm", ``
+  !mmap_w a. bir_load_mmap_w (mmap_w: 'a word |-> 'b word) a =
+	       case FLOOKUP mmap_w a of
+		 | NONE => 0w
+		 | SOME v => v
+``,
+  REPEAT STRIP_TAC >>
+  REWRITE_TAC [bir_load_mmap_w_def] >>
+
+  REWRITE_TAC [bir_load_mmap_def, bir_mmap_w_w2n_def] >>
+
+  Cases_on `FLOOKUP mmap_w a` >- (
+    `~((w2n a) IN (IMAGE w2n (FDOM mmap_w)))` by (
+      FULL_SIMP_TAC (std_ss++pred_setSimps.PRED_SET_ss) [FLOOKUP_DEF] >>
+      METIS_TAC [w2n_11]
+    ) >>
+    ASM_SIMP_TAC (std_ss++pred_setSimps.PRED_SET_ss) [FLOOKUP_FUN_FMAP, FDOM_FINITE, IMAGE_FINITE]
+  ) >>
+
+  `((w2n a) IN (IMAGE w2n (FDOM mmap_w)))` by (
+    FULL_SIMP_TAC (std_ss++pred_setSimps.PRED_SET_ss) [FLOOKUP_DEF]
+  ) >>
+  `a IN (FDOM mmap_w)` by (
+    FULL_SIMP_TAC (std_ss++pred_setSimps.PRED_SET_ss) [w2n_11]
+  ) >>
+  FULL_SIMP_TAC (std_ss++pred_setSimps.PRED_SET_ss) [FLOOKUP_FUN_FMAP, FDOM_FINITE, IMAGE_FINITE, n2w_w2n, FLOOKUP_DEF]
+);
+
+
+val bir_mf2mm_def = Define `
+   (bir_mf2mm (mf: 'a word -> 'b word)) : 'a word |-> 'b word =
+      FUN_FMAP mf (UNIV:'a word -> bool)(*{ x | mf x <> 0w }*)
+    `;
+
+
+
+val bir_FLOOKUP_mmap_w2n__thm = store_thm("bir_FLOOKUP_mmap_w2n__thm", ``
+  !mmap_w a. FLOOKUP (bir_mmap_w_w2n mmap_w) (a MOD dimword (:'a)) = OPTION_MAP w2n (FLOOKUP mmap_w (n2w a : 'a word))
```

Comment:
Can you break these long lines? :grin: (this one and all the long definitions below)